File: sdk/java/src/main/java/org/finos/openresourcebroker/sdk/sse/SseReader.java
// Layer 5: SSE Reader + Reconnect
package org.finos.openresourcebroker.sdk.sse;

import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * Parses SSE (Server-Sent Events) frames from an InputStream.
 *
 * <p>SSE wire format used by ORB:
 * <pre>
 *   data: &lt;json&gt;\n\n          — normal event
 *   data: {}\n\n              — terminal sentinel (stream done)
 * </pre>
 *
 * <p>Only {@code data:} lines are processed; {@code event:}, {@code id:},
 * and {@code retry:} are accepted but ignored (ORB does not use them).
 *
 * <p>The frame buffer is bounded at {@link #MAX_SSE_FRAME_BYTES} (4 MiB, matching
 * the Go/TS/Kotlin/C# SDKs).  A server that never emits a newline or emits an
 * enormous {@code data:} line would otherwise grow client memory without bound;
 * on overflow the reader throws a terminal {@link IOException} so the caller can
 * abort rather than reconnect and spin on the same oversized frame.
 */
public class SseReader implements Closeable {

    /** Maximum size of a single accumulated SSE frame (4 MiB). */
    public static final int MAX_SSE_FRAME_BYTES = 4 * 1024 * 1024;

    private final InputStream stream;

    public SseReader(InputStream stream) {
        this.stream = stream;
    }

    /**
     * Read the next complete SSE frame.
     *
     * @return the next frame, or {@code null} when the stream ends normally
     * @throws IOException if an I/O error occurs or a frame exceeds
     *                     {@link #MAX_SSE_FRAME_BYTES}
     */
    public SseFrame next() throws IOException {
        List<String> dataLines = new ArrayList<>();
        long frameBytes = 0;

        String line;
        while ((line = readLine()) != null) {
            if (line.isEmpty()) {
                // Blank line = end of event
                if (dataLines.isEmpty()) {
                    continue; // heartbeat / keep-alive blank line
                }
                String data = String.join("\n", dataLines);
                dataLines.clear();
                return new SseFrame(data);
            }

            if (line.startsWith("data:")) {
                String val = line.substring(5);
                if (!val.isEmpty() && val.charAt(0) == ' ') {
                    val = val.substring(1);
                }
                frameBytes += val.length() + 1; // +1 for the joining newline
                if (frameBytes > MAX_SSE_FRAME_BYTES) {
                    throw new IOException("SSE frame exceeds maximum allowed "
                            + MAX_SSE_FRAME_BYTES + " bytes");
                }
                dataLines.add(val);
            }
            // Ignore event:, id:, retry: lines
        }

        // EOF
        return null;
    }

    /**
     * Read a single line, bounded by {@link #MAX_SSE_FRAME_BYTES}, terminated by
     * LF (CR is stripped).  Reads raw bytes so an unbounded, newline-free line
     * cannot exhaust the heap via a decoding buffer.
     *
     * @return the line without its terminator, or {@code null} at EOF before any
     *         byte was read
     */
    private String readLine() throws IOException {
        java.io.ByteArrayOutputStream buf = new java.io.ByteArrayOutputStream(256);
        int b;
        boolean any = false;
        while ((b = stream.read()) != -1) {
            any = true;
            if (b == '\n') {
                return decode(buf);
            }
            if (b == '\r') {
                continue; // strip CR; the following LF ends the line
            }
            if (buf.size() >= MAX_SSE_FRAME_BYTES) {
                throw new IOException("SSE line exceeds maximum allowed "
                        + MAX_SSE_FRAME_BYTES + " bytes");
            }
            buf.write(b);
        }
        if (!any && buf.size() == 0) return null;
        return decode(buf);
    }

    private static String decode(java.io.ByteArrayOutputStream buf) {
        return buf.toString(StandardCharsets.UTF_8);
    }

    @Override
    public void close() {
        if (stream == null) return;
        try {
            stream.close();
        } catch (IOException ignored) {}
    }
}

File: sdk/kotlin/src/main/kotlin/org/finos/openresourcebroker/sdk/sse/SseReader.kt
/**
 * Layer 5: Server-Sent Events Reader + Reconnect
 *
 * Parses SSE wire format from an InputStream, emits typed [SseFrame] objects,
 * and reconnects with exponential back-off when the connection is dropped.
 *
 * ORB SSE wire format:
 *   data: <json>\n\n   — normal event
 *   data: {}\n\n       — terminal sentinel (stream done)
 *
 * Kotlin Flow is used as the async primitive so callers can collect with
 * structured concurrency.
 */

package org.finos.openresourcebroker.sdk.sse

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import org.finos.openresourcebroker.sdk.client.OrbApiError
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import kotlin.math.min

/**
 * Maximum size of a single SSE line / accumulated frame (4 MiB, matching the
 * Go/TS/Java/C# SDKs). A server that never emits a newline, or emits an enormous
 * `data:` line, would otherwise grow the client heap without bound. On overflow
 * the reader throws so the stream terminates rather than reconnecting and
 * spinning on the same oversized frame.
 */
internal const val MAX_SSE_FRAME_BYTES = 4 * 1024 * 1024

// ---------------------------------------------------------------------------
// Wire types
// ---------------------------------------------------------------------------

data class SseFrame(
    val data: String,
    val event: String? = null,
    val id: String? = null,
    val retry: Long? = null,
)

// ---------------------------------------------------------------------------
// ORB-specific payload types
// ---------------------------------------------------------------------------

data class OrbMachine(
    val machine_id: String = "",
    val name: String? = null,
    val status: String? = null,
    val result: String? = null,
    val private_ip: String? = null,
    val public_ip: String? = null,
    val launch_time: String? = null,
    val message: String? = null,
)

data class OrbSseRequest(
    val request_id: String = "",
    val status: String = "",
    val message: String? = null,
    val requested_count: Int? = null,
    val successful_count: Int? = null,
    val failed_count: Int? = null,
    val machines: List<OrbMachine>? = null,
)

data class OrbSsePayload(
    val requests: List<OrbSseRequest>? = null,
    val event_type: String? = null,
    val sentinel: Boolean? = null,
)

val TERMINAL_STATUSES = setOf(
    "complete", "completed", "failed", "error", "cancelled", "canceled", "partial", "timeout"
)

/**
 * True if this frame is the ORB terminal sentinel (data: {}).
 */
fun SseFrame.isSentinel(): Boolean = data.trim() == "{}"

// ---------------------------------------------------------------------------
// Frame parsing
// ---------------------------------------------------------------------------

/**
 * Read a single LF-terminated line from a raw byte [stream], stripping a trailing
 * CR. Reads raw bytes (not a decoding [java.io.BufferedReader]) and caps the line
 * at [MAX_SSE_FRAME_BYTES] so an unbounded, newline-free line cannot exhaust the
 * heap.
 *
 * @return the decoded line without its terminator, or `null` at EOF before any
 *   byte was read
 * @throws IOException if the line exceeds [MAX_SSE_FRAME_BYTES]
 */
private fun readBoundedLine(stream: InputStream): String? {
    val buf = ByteArrayOutputStream(256)
    var any = false
    while (true) {
        val b = stream.read()
        if (b == -1) break
        any = true
        if (b == '\n'.code) return buf.toString(Charsets.UTF_8.name())
        if (b == '\r'.code) continue // strip CR; the following LF ends the line
        if (buf.size() >= MAX_SSE_FRAME_BYTES) {
            throw IOException("SSE line exceeds maximum allowed $MAX_SSE_FRAME_BYTES bytes")
        }
        buf.write(b)
    }
    if (!any) return null
    return buf.toString(Charsets.UTF_8.name())
}

/**
 * Parse SSE frames from a blocking [InputStream].
 *
 * Runs the blocking reads on [Dispatchers.IO]. Because a blocking socket read
 * cannot be interrupted by coroutine cancellation, an [invokeOnCompletion]
 * handler closes the underlying [stream] when the collecting scope is cancelled
 * (or the flow otherwise completes), unblocking the in-progress read with an
 * IOException so the IO thread and UDS connection are released promptly.
 */
internal fun parseSseFrames(stream: InputStream): Flow<SseFrame> = flow {
    var currentData: StringBuilder? = null
    var frameBytes = 0
    var currentEvent: String? = null
    var currentId: String? = null
    var currentRetry: Long? = null

    // Close the stream on cancellation/completion so a blocking readBoundedLine
    // parked on the socket is released instead of leaking the IO thread.
    val job = currentCoroutineContext()[Job]
    val handle = job?.invokeOnCompletion {
        try { stream.close() } catch (_: Exception) {}
    }

    try {
        while (true) {
            val line = readBoundedLine(stream) ?: break

            if (line.isEmpty()) {
                // Blank line = dispatch event
                val data = currentData
                if (data != null) {
                    emit(
                        SseFrame(
                            data = data.toString(),
                            event = currentEvent,
                            id = currentId,
                            retry = currentRetry,
                        )
                    )
                }
                currentData = null
                frameBytes = 0
                currentEvent = null
                currentRetry = null
                // Note: currentId persists across frames (Last-Event-ID)
                continue
            }

            val colonIdx = line.indexOf(':')
            if (colonIdx == -1) {
                // Line is a field name with no value
                continue
            }

            val field = line.substring(0, colonIdx)
            val rawValue = line.substring(colonIdx + 1)
            val value = if (rawValue.startsWith(" ")) rawValue.substring(1) else rawValue

            when (field) {
                "data" -> {
                    frameBytes += value.length + 1 // +1 for the joining newline
                    if (frameBytes > MAX_SSE_FRAME_BYTES) {
                        throw IOException(
                            "SSE frame exceeds maximum allowed $MAX_SSE_FRAME_BYTES bytes"
                        )
                    }
                    if (currentData == null) {
                        currentData = StringBuilder(value)
                    } else {
                        currentData.append('\n').append(value)
                    }
                }
                "event" -> currentEvent = value
                "id" -> currentId = value
                "retry" -> currentRetry = value.toLongOrNull()
                "" -> { /* comment, ignore */ }
            }
        }
    } finally {
        handle?.dispose()
    }
}.flowOn(Dispatchers.IO)

// ---------------------------------------------------------------------------
// Reconnecting SSE stream
// ---------------------------------------------------------------------------

data class SseStreamOptions(
    val initialDelayMs: Long = 1_000L,
    val maxDelayMs: Long = 30_000L,
)

/**
 * Reconnecting SSE stream as a Kotlin [Flow].
 *
 * @param connect Called on each (re)connection attempt. Receives the last
 *   event ID for the Last-Event-ID header. Must return an InputStream from
 *   which SSE bytes are read.
 * @param opts Reconnect policy configuration.
 */
fun sseStream(
    connect: suspend (lastEventId: String?) -> InputStream,
    opts: SseStreamOptions = SseStreamOptions(),
): Flow<SseFrame> = flow {
    var delayMs = opts.initialDelayMs
    var lastEventId: String? = null

    while (true) {
        val stream: InputStream = try {
            connect(lastEventId)
        } catch (e: OrbApiError) {
            // 4xx errors are terminal — do not reconnect (no frame-parse, no retry).
            // 5xx is a candidate for reconnect with back-off.
            if (e.statusCode in 400..499) throw e
            delay(jitter(delayMs))
            delayMs = min(delayMs * 2, opts.maxDelayMs)
            continue
        } catch (e: Exception) {
            if (e is CancellationException) throw e
            delay(jitter(delayMs))
            delayMs = min(delayMs * 2, opts.maxDelayMs)
            continue
        }

        var gotEvents = false
        try {
            parseSseFrames(stream).collect { frame ->
                if (frame.id != null) lastEventId = frame.id
                if (frame.retry != null) delayMs = frame.retry
                gotEvents = true
                emit(frame)
                if (frame.isSentinel()) {
                    // Terminal sentinel — stop the flow cleanly
                    throw TerminalSentinelException()
                }
            }
        } catch (e: TerminalSentinelException) {
            return@flow
        } catch (e: CancellationException) {
            throw e
        } catch (_: Exception) {
            // Stream error — reconnect below
        } finally {
            try { stream.close() } catch (_: Exception) {}
        }

        // Clean stream end or error — reconnect with back-off
        if (gotEvents) {
            delayMs = opts.initialDelayMs // reset after successful connection
        }

        delay(jitter(delayMs))
        delayMs = min(delayMs * 2, opts.maxDelayMs)
    }
}

private fun jitter(delayMs: Long): Long {
    return (delayMs * (0.5 + Math.random() * 0.5)).toLong()
}

/** Thrown internally to terminate the reconnect loop on sentinel. */
internal class TerminalSentinelException : Exception()
